File: src/main/java/io/fobo66/crypto/Lab1.java
package io.fobo66.crypto;

import org.apache.commons.cli.*;

import javax.xml.bind.DatatypeConverter;

public class Lab1 {

    public static void main(String[] args) {
        String clearText = "Hello World!";
        String key = "password";
        DESMode mode = DESMode.ECB;

        Options options = new Options();
        options.addOption("file", true, "Input file");
        options.addOption("key", true, "Encryption key");
        options.addOption("m", "mode", true, "Operation mode for DES (ECB, CBC, CFB, OFB)");

        CommandLineParser parser = new DefaultParser();
        try {
            CommandLine cmd = parser.parse(options, args);

            if (cmd.hasOption("file")) {
                clearText = loadClearTextFromFile(cmd.getOptionValue("file"));
            }

            if (cmd.hasOption("key")) {
                key = cmd.getOptionValue("key");
            }

            if (cmd.hasOption("m")) {
                mode = loadDESMode(cmd);
            }

            byte[] encryptedText = DES.encrypt(clearText.getBytes(), key.getBytes(), mode);
            byte[] decryptedText = DES.decrypt(encryptedText, key.getBytes(), mode);

            printResults(clearText, encryptedText, decryptedText);
        } catch (ParseException e) {
            System.err.println("Parsing failed. Reason: " + e.getMessage());
        }
    }

    private static String loadClearTextFromFile(String filePath) {
        return "";
    }

    private static DESMode loadDESMode(CommandLine cmd) {
        return DESMode.valueOf(cmd.getOptionValue("m").toUpperCase());
    }

    private static void printResults(String clearText, byte[] encryptedText, byte[] decryptedText) {
        System.out.println("Clear text: " + clearText);
        System.out.println("Encrypted text: " + DatatypeConverter.printHexBinary(encryptedText));
        System.out.println("Decrypted text: " + new String(decryptedText));
    }
}

File: src/main/java/io/fobo66/crypto/Lab1.kt
package io.fobo66.crypto

import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.core.main
import com.github.ajalt.clikt.parameters.options.default
import com.github.ajalt.clikt.parameters.options.help
import com.github.ajalt.clikt.parameters.options.option
import com.github.ajalt.clikt.parameters.transform.theme
import com.github.ajalt.clikt.parameters.types.choice
import com.github.ajalt.clikt.parameters.types.path
import java.io.IOException
import java.math.BigInteger
import kotlin.io.path.readText

const val RADIX = 16

class Lab1 : CliktCommand() {
    val inputFile by option("-f", "--file")
        .path(mustBeReadable = true, canBeDir = false, mustExist = true)
        .help { theme.info("Input file") }

    val key by option("-k", "--key")
        .help { theme.info("Encryption key") }
        .default("password")

    val mode by option("-m", "--mode")
        .choice("ECB" to DESMode.ECB, "CBC" to DESMode.CBC, "CFB" to DESMode.CFB, "OFB" to DESMode.OFB)
        .help { theme.info("Operation mode for DES (ECB, CBC, CFB, OFB)") }
        .default(DESMode.ECB)

    override fun run() {
        var clearText = "Hello World!"

        try {
            if (inputFile != null) {
                echo("Reading cleartext from file $inputFile...")
                clearText = inputFile?.readText().orEmpty()
            }
            echo("Using $mode mode of operation...")
            val encryptedText = DES.encrypt(clearText.toByteArray(), key.toByteArray(), mode)
            val decryptedText = DES.decrypt(encryptedText, key.toByteArray(), mode)
            printResults(clearText, encryptedText, decryptedText)
        } catch (e: IOException) {
            throw IllegalStateException("Failed to read cleartext from file", e)
        }
    }

    private fun printResults(clearText: String, encryptedText: ByteArray, decryptedText: ByteArray) {
        echo("Clear text: $clearText")
        echo("Encrypted text: ${BigInteger(1, encryptedText).toString(RADIX)}")
        echo("Decrypted text: ${decryptedText.toString(Charsets.UTF_8)}")
    }

}

fun main(args: Array<String>) = Lab1().main(args)
